File: src/main/java/com/machfour/macros/core/Serving.java
package com.machfour.macros.core;

import com.machfour.macros.data.ColumnData;
import com.machfour.macros.data.Table;
import com.machfour.macros.data.Schema;
import com.sun.istack.internal.NotNull;

public class Serving extends MacrosEntity<Serving> {

    private Food food;
    private QuantityUnit quantityUnit;

    public Serving(ColumnData<Serving> data, ObjectSource objectSource) {
        super(data, objectSource);
        food = null;
        quantityUnit = null;
    }

    @Override
    public Table<Serving> getTable() {
        return Schema.ServingTable.instance();
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Serving && super.equals(o);
    }

    public Food getFood() {
        return food;
    }

    public void setFood(@NotNull Food f) {
        assert food == null && foreignKeyMatches(this, Schema.ServingTable.FOOD_ID, f);
        food = f;
    }

    @NotNull
    public String getQuantityUnitAbbr() {
        return getData(Schema.ServingTable.QUANTITY_UNIT);
    }

    public QuantityUnit getQuantityUnit() {
        return quantityUnit;
    }

    public void setQuantityUnit(@NotNull QuantityUnit q) {
        assert getQuantityUnit() == null && foreignKeyMatches(this, Schema.ServingTable.QUANTITY_UNIT, q);
        quantityUnit = q;
    }

    @NotNull
    public Long getFoodId() {
        return getData(Schema.ServingTable.FOOD_ID);
    }

    @NotNull
    public String getName() {
        return getData(Schema.ServingTable.NAME);
    }

    @NotNull
    public Double getQuantity() {
        return getData(Schema.ServingTable.QUANTITY);
    }

    @NotNull
    public Boolean isDefault() {
        return getData(Schema.ServingTable.IS_DEFAULT);
    }
}

File: src/main/kotlin/com/machfour/macros/entities/Serving.kt
package com.machfour.macros.entities

import com.machfour.macros.core.Factory
import com.machfour.macros.core.MacrosEntityImpl
import com.machfour.macros.core.ObjectSource
import com.machfour.macros.core.PortionMeasurement
import com.machfour.macros.entities.auxiliary.Factories
import com.machfour.macros.nutrients.QUANTITY
import com.machfour.macros.schema.ServingTable
import com.machfour.macros.sql.RowData
import com.machfour.macros.sql.Table
import com.machfour.macros.units.unitWithAbbr

class Serving internal constructor(data: RowData<Serving>, objectSource: ObjectSource) :
    MacrosEntityImpl<Serving>(data, objectSource), PortionMeasurement {

    companion object {
        // factory has to come before table if it's an instance variable
        val factory: Factory<Serving>
            get() = Factories.serving
    }

    val qtyUnitAbbr: String
        get() = data[ServingTable.QUANTITY_UNIT]!!

    val qtyUnit = unitWithAbbr(qtyUnitAbbr)

    init {
        check(QUANTITY.compatibleWith(qtyUnit)) { "Invalid unit $qtyUnit for nutrient $QUANTITY" }
    }

    lateinit var food: Food
        private set

    override val factory: Factory<Serving>
        get() = Companion.factory
    override val table: Table<Serving>
        get() = ServingTable

    val foodId: Long
        get() = data[ServingTable.FOOD_ID]!!

    val quantity: Double
        get() = data[ServingTable.QUANTITY]!!

    val isDefault: Boolean
        get() = data[ServingTable.IS_DEFAULT]!!

    val notes: String?
        get() = data[ServingTable.NOTES]

    override fun equals(other: Any?): Boolean {
        return other is Serving && super.equals(other)
    }

    override fun hashCode(): Int {
        return super.hashCode()
    }

    fun initFood(f: Food) {
        assert(foreignKeyMatches(this, ServingTable.FOOD_ID, f))
        food = f
    }

    override val name: String
        get() = data[ServingTable.NAME]!!


    // Measurement functions
    override val unitMultiplier = quantity

    override val baseUnit
        get() = qtyUnit

    override val isVolumeMeasurement
        get() = qtyUnit.isVolumeMeasurement

}
